proposals/operator-ingress-sketch.md

```diff
@@ -0,0 +1,189 @@
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: Proxy
+metadata:
+  name: myproxy
+spec:
+   # All fields are optional - an empty proxy is sufficient to get going
+
+   infrastructure:
+      replics: 2
+      # resources / env
+      # labels / annotations to be applied.
+
+   # Refereneces one or more ingresses.
+   # If no ingresses are defined, the operator will behave as if a plain clusterip ingress is defined called default.
+   ingresses:
+   - myclusterip
+   - foobar
+status:
+   # reports ingresses
+```
+
+# ProxyIngress
+
+Owned by the Infrastructure admin.
+
+Declares how the proxy exposes itselt within the cluster or to the outside world.
+
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: ProxyIngress
+metadata:
+  name: myclusterip
+spec:
+  # oneOf
+  service:
+    type: ClusterIP|LoadBalancer
+    brokerAddressPattern: $(node)-${vitrtualClusterName}.example.com  (LoadBalancer only)
+    protocol: TCP|TLS (ClusterIP supports both, LoadBalancer TLS only)
+    port: # optional, defaults to 9082
+  openShiftRoute:
+    ingressControllName: # ingress controller name (optional)
+  gateway:
+    type: TLSRoute
+    # think about this
+```      
+
+# VirtualCluster
+
+Created by a developer
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: VirtualCluster
+metadata:
+  name: mycluster
+spec:
+  proxyReference:
+    kind: ProxyIngress
+    group: proxy.kroxylicious.io
+    name: myproxy
+ 
+  # list of ingresses the virtual cluster will use.  If the ingresses object is omitted, the virtual cluster uses all
+  # ingress declared at by the proxy.  If the proxy is using the default ingress, that is what the virtualcluster will
+  # use too.
+  ingresses:
+  - name: myclusterip
+
+  # Points to the cluster being proxied.  Can either be a external service or a Strimzi Kafka resource.
+  targetClusterRef:
+    kind: Kafka|Service
+    group: strimzi.io|<empty>
+    name: name
+    listenerName: <strimzi listener name>  (not allowed if kind is Service)
+    protocol: TCP|TLS (not allowed if kind is Kafka)
+    # client cert and trust
+
+  filters:
+    ... 
+```
+
+# Worked examples
+
+## Virtual cluster exposed on cluster using a cluster IP.  This exploits the defaulting.
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: Proxy
+metadata:
+  name: myproxy
+spec: {}
+status:
+  ingresses:
+  - name: default
+    service:
+      type: ClusterIP
+      protocol: TCP
+```
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: VirtualCluster
+metadata:
+  name: mycluster
+spec:
+  proxyReference:
+    name: myproxy
+    kind: ProxyIngress
+    group: proxy.kroxylicious.io
+
+  filters:
+  - xxxx: ...
+ 
+  targetClusterRef:
+    name: my-cluster
+    group: strimzi.io|<empty>
+    kind: strimzi
+    listener: my-strimzi-listener
+```
+
+What would operator create:
+* kroxylicious deployment, 1 replica
+* n+1 ClusterIP services (1 per broker + 1 bootstrap).  It would watch the strimzi resource in order to know n
```

Comment:
> It would watch the strimzi resource in order to know n

If we're going to go this route then ading this functionality to the proxy is something we can progress in parallel to bikeshedding on this API.

Reply:
You mean have the proxy probe the target cluster and discover the broker topology and somehow have the operator discover that from the proxy? REST API?  I don't think that's completely straightforward - I see a cold start issue.

1. Cluster gets started (3 brokers + bootstrap route)
2. Proxy gets started.  Proxy doesn't know how many brokers cluster has.
3. Operator establishes proxy bootstrap route.
4. Client connects through proxy to bootstrap, proxy forwards connection.
5. Proxy learns cluster has 3 brokers.
6. Operator learns (polling?) the Proxy that the Cluster has three brokers.  
7. Operator establishes broker routes.

Until we've reached step 7, we'll see connection failures.